internal/state/graph/httproute.go

```diff
@@ -326,40 +320,49 @@ func bindRouteToListeners(r *Route, gw *Gateway) {
 		// FIXME(pleshakov)
 		// For now, let's do simple matching.
 		// However, we need to also support wildcard matching.
-		// More over, we need to handle cases when a Route host matches multiple HTTP listeners on the same port
-		// when sectionName is empty and only choose one listener.
-		// For example:
-		// - Route with host foo.example.com;
-		// - listener 1 for port 80 with hostname foo.example.com
-		// - listener 2 for port 80 with hostname *.example.com;
-		// In this case, the Route host foo.example.com should choose listener 1, as it is a more specific match.
-
-		l, exists := gw.Listeners[string(*routeRef.SectionName)]
-		if !exists {
-			// FIXME(pleshakov): Add a proper condition once it is available in the Gateway API.
-			// https://github.com/nginxinc/nginx-kubernetes-gateway/issues/306
-			attachment.FailedCondition = conditions.NewTODO("listener is not found")
-			continue
+
+		bind := func(l *Listener) (valid bool) {
+			if !l.Valid {
+				return false
+			}
+
+			hostnames := findAcceptedHostnames(l.Source.Hostname, r.Source.Spec.Hostnames)
+			if len(hostnames) == 0 {
+				return true // listener is valid, but return without attaching due to no matching hostnames
+			}
+
+			attachment.Attached = true
+			for _, h := range hostnames {
+				l.AcceptedHostnames[h] = struct{}{}
+			}
+			l.Routes[client.ObjectKeyFromObject(r.Source)] = r
+
+			return true
 		}
 
-		if !l.Valid {
+		var valid bool
+		if getSectionName(routeRef.SectionName) == "" {
+			for _, l := range gw.Listeners {
```

Comment:
would it make sense to add a test when we bind with an empty section name and no valid listeners are there?

Reply:
What's the test coverage look like for this function? It's getting pretty complicated and hard to read. I wonder if it makes sense to try and break it into different functions.